### PR TITLE
WIP: Always use np.intp for sparse matrix indices

### PR DESCRIPTION
**Pull request by eric-wieser**
Testing something brought up by scikit-learn/scikit-learn#11295. Right now, scipy seems to allow sparse matrices that have indices which are too large to be made dense on 32-bit systems, which may or may not be a feature.

### CODE REVIEW

**Comment by tylerjereddy** [comment]
#9087 may be related and is soon to be merged I think; even if it is not exactly the same problem that "upcasting" should probably be merged first I suppose.

**Comment by perimosocordiae** [comment]
@eric-wieser Can you summarize the issue here? I'm not sure I understand the context for this change.

**Comment by pv** [comment]
It's unclear to me what this PR attempts --- the problem that it tries to solve is not explained, and remains unclear after reading the scikit-learn thread.

It is a feature that the indices may be either 32-bit or 64-bit, regardless of whether the platform is 32-bit or 64-bit. The default index size is always 32-bit, both from memory conservation issues and also due to backward compatibility issues. It is not possible to change this.

SuperLU indeed does not work with 64-bit indices (even on 64-bit platforms), so changing to intp won't help.

**Comment by eric-wieser** [comment]
>the problem that it tries to solve is not explained

Apologies for this - in many ways, this was just a case of "let's run it again the CI on some different platforms and see what happens".
